### ISSUE TITLE
Cannot open non-empty directory/file unless it contains at least one test case

### ISSUE THREAD

**Issue by yanne** [robotframework/RIDE#127]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92) by c.martin.taylor on 12 Feb 2009*

When starting a new project I usually want to create a test suite folder
structure with only an __init__.html in each directory.  Then I create the
test suite HTML files with only a documentation line that says what it will
contain.  If you do this, either manually or using RIDE, and then try to
re-open the directory with RIDE it complains that it is not a valid
directory.  The only work-around I've found for this is to put at least one
"dummy" test into each .html file.  Then RIDE will recognize and load the file.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c1) by @yanne on 12 Feb 2009*


This behavior is due to the fact that we use RF modules for parsing, and RF ignores
test case files that do not contain any test cases. To fix this, we would need either

1) change the RF parsing API to somehow accommodate this, or
2) implement our own parser

1) is probably much easier but requires change in RF, so this is possible to only
after RF 2.1 is released. Since 2) requires much more work, waiting for 1) is
probably as fast anyway.

Pekka, any comments on this?

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c2) by @pekkaklarck on 16 Feb 2009*


As Janne wrote, this problem is due to RF parsing modules ignoring such
files/directories. I just submitted and issue [1] to RF about making parsing more
configurable. Fixing this on RIDE side without those changes would be so complicated
that we probably need to wait them before we can resolve this fully.

Special casing empty directories would be pretty simple even earlier, though. We can
just do os.listdir and if it returns an empty list create an empty suite with a name
got from the directory name. If this minor enhancement is considered important, a
separate issue can be opened for it.


[1] <a href="https://code.google.com/p/robotframework/issues/detail?id=232" rel="nofollow">https://code.google.com/p/robotframework/issues/detail?id=232</a>

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c3) by bryan.oakley on 29 Jul 2009*


Since pekka suggested opening a new issue to special-case empty directories, I've
done so. See <a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=202" title="Possibility to open an empty directory or file"> issue #202 </a>

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c5) by @yanne on 22 Sep 2009*


Supporting this requires changes to Robot Framework's parsing modules. These are
planned for RF 2.2

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c6) by @pekkaklarck on 9 Oct 2009*


After <a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=202" title="Possibility to open an empty directory or file"> issue 202 </a> implemented RIDE can open empty directories and files but doesn't
support them inside directories.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c7) by c.martin.taylor on 5 Jan 2010*


<a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=359" title="Starting RIDE in an empty folder doesn't work"> Issue 359 </a> is actually an instance of this issue.  Why would the existence of a .bat
file in a directory cause this problem.  Surely RF should only try to parse file
types it supports such as .tsv and .html files?

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c8) by @pekkaklarck on 5 Jan 2010*


<a class="closed_ref" href="https://code.google.com/p/robotframework-ride/issues/detail?id=359" title="Starting RIDE in an empty folder doesn't work"> Issue 359 </a> has been merged into this issue.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c9) by @pekkaklarck on 5 Jan 2010*


RF doesn't parse `.bat` files. The reason they prevent opening the otherwise empty
directory is that totally empty dirs are special cased on Ride itself.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c11) by dragonfyre13 on 26 Jan 2012*


There's another fix, for TSV at least. I ran across this just today, and found out that if I put:
"""
*Test Case*	*Action*	*Action*	*Action*
"""

at the top of the file, even if there's no test case underneath it, the file opens fine in the latest version of RIDE (with the latest RF). Maybe it's as easy as RIDE adding that header in each file regardless of if a test case has been defined yet?

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c12) by @mkorpela on 11 Sep 2012*


Thanks dragonfyre13!

Sounds like a plan!

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c13) by @mkorpela on 11 Sep 2012*


Should be fixed now.

What I have done is to mark new suite files dirty at the beginning - this makes it possible to save them without modifications. An empty suite file can be opened in RIDE.

<a href="https://github.com/robotframework/RIDE/commit/47addea7c088afcb04270c65ba8d1123955814cc" rel="nofollow">https://github.com/robotframework/RIDE/commit/47addea7c088afcb04270c65ba8d1123955814cc</a>

I'll add some test cases so that this will work also in the future.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c14) by @mkorpela on 11 Sep 2012*


Suite directory without test cases will be reloaded as resource directory - this is not a really nice thing.

I think that the best way to handle this problem is to remove the concept of a resource directory and just think them as suite directories.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c15) by @mkorpela on 11 Sep 2012*


Now resource folders are also suite folders: 
<a href="https://github.com/robotframework/RIDE/commit/0725265f1eb8ed0b9b9ec753b90f8b3feb5b1b6a" rel="nofollow">https://github.com/robotframework/RIDE/commit/0725265f1eb8ed0b9b9ec753b90f8b3feb5b1b6a</a>

I'm still going to remove DirectoryController class as it should not have any reason to exist anymore.

**Comment by yanne** [comment]
> *Originally submitted to [Google Code](http://code.google.com/p/robotframework-ride/issues/detail?id=92#c16) by @mkorpela on 12 Sep 2012*


DirectoryController removed:
<a href="https://github.com/robotframework/RIDE/commit/c93844af0cd05997a46a5fba8a10da1e2c1bcecb" rel="nofollow">https://github.com/robotframework/RIDE/commit/c93844af0cd05997a46a5fba8a10da1e2c1bcecb</a>

This part of code is using multiple inheritance so this could have some ugly side-effects.
